### PR TITLE
Alternate \page{curlies}

### PR DESCRIPTION
**Pull request by calculuschild**
## Description

Takes #3899 and moves the handling of curlies back into Markdown.js. Small adjustments to regex, etc.

Opening as a new PR as I was getting too many merge conflicts to easily track. Note that these changes *only* apply to V3. Legacy should be unaffected.

##### Usage:
Similar to an inline single mustache.

`\page{wide,banna=fail,color:black}`

## Related Issues or Discussions

Solves https://github.com/naturalcrit/homebrewery/issues/3901

*Reviewers, refer to this list when testing features, or suggest new items*
- [ ] Verify new features are functional
  - [ ] Adding a style includes a style on the `<div class='page'>` instance.
  - [ ] Adding a class includes a class on the `<div class='page'>` instance.
  - [ ] Adding an HTML Attribute adds the attribute on the `<div class='page'>` instance.
  - [ ] Adding an ID does ***not*** add it to the `<div class='page'>` instance.
  - [ ] Mixed forms of above work as expected
  - [ ] CSS styles with hyphens (e.g. `line-height`) work (Page is a React component, which requires camelCase input)
  - [ ] Curlies mix properly with other pages styles (e.g. the page "shadow" option from the toolbar)
  - [ ] `\page` on the first line of a document will *not* create a page break, and is counted as page 1
- [ ] Verify old features have not broken
  - [ ] `\page` works without values.
  - [ ] Highlighting of `\page` still works
  - [ ] Page number display is accurate
  - [ ] Jumping to source/brew still works in every case (including Legacy)
- [ ] Test for edge cases / try to break things
  - [ ] Ensure CSS Properties ( --FOO ) can be set.
- [ ] Identify opportunities for simplification and refactoring
- [ ] Check for code legibility and appropriate comments

<details><summary>Copy the list here</summary>
<p>

```

- [ ] Verify new features are functional
  - [ ] Adding a style includes a style on the `<div class='page'>` instance.
  - [ ] Adding a class includes a class on the `<div class='page'>` instance.
  - [ ] Adding an HTML Attribute adds the attribute on the `<div class='page'>` instance.
  - [ ] Adding an ID does ***not*** add it to the `<div class='page'>` instance.
  - [ ] Mixed forms of above work as expected
  - [ ] CSS styles with hyphens (e.g. `line-height`) work (Page is a React component, which requires camelCase input)
  - [ ] Curlies mix properly with other pages styles (e.g. the page "shadow" option from the toolbar)
  - [ ] `\page` on the first line of a document will *not* create a page break, and is counted as page 1
- [ ] Verify old features have not broken
  - [ ] `\page` works without values.
  - [ ] Highlighting of `\page` still works
  - [ ] Page number display is accurate
  - [ ] Jumping to source/brew still works in every case (including Legacy)
- [ ] Test for edge cases / try to break things
  - [ ] Ensure CSS Properties ( --FOO ) can be set.
- [ ] Identify opportunities for simplification and refactoring
- [ ] Check for code legibility and appropriate comments

```

</p>
</details>

### CODE REVIEW

**Comment by 5e-Cleric** [comment]
Using an empty inject breaks the syntax, which doesn't make much sense, allow for `\page{}`

File: client/homebrew/brewRenderer/brewRenderer.jsx [hunk 1]
```diff
@@ -40,7 +40,7 @@ const BrewPage = (props)=>{
 		...props
 	};
 	const pageRef = useRef(null);
-	const cleanText = safeHTML(props.contents);
+	let cleanText = safeHTML(props.contents);
```
Comment (calculuschild) [inline client/homebrew/brewRenderer/brewRenderer.jsx:43]: Should stay as `const`. Probably leftover from my rearranging things

**Comment by calculuschild** [comment]
> Using an empty inject breaks the syntax, which doesn't make much sense, allow for `\page{}`

Fixed.

**Comment by calculuschild** [comment]
Any final comments, @5e-Cleric , @dbolack-ab , @ericscheid ?

**Comment by 5e-Cleric** [comment]
Right now, when setting style for the *first* page, spaces are being allowed between page break and style inject, but not for other pages. Additionally, these spaces, while they don't break the regex for the inject, do break the syntax highlight.

**Comment by 5e-Cleric** [comment]
Also, class or id injection doesn't work for either ( `\page{#myPageId,myPageClass}`

**Comment by 5e-Cleric** [comment]
Will update these as i go
- [ ] Verify new features are functional
  - [ ] Adding a style includes a style on the `<div class='page'>` instance.
  - [ ] Adding a class includes a class on the `<div class='page'>` instance.
  - [ ] Adding an HTML Attribute adds the attribute on the `<div class='page'>` instance.
  - [ ] Adding an ID does ***not*** add it to the `<div class='page'>` instance.
  - [ ] Mixed forms of above work as expected
  - [ ] CSS styles with hyphens (e.g. `line-height`) work (Page is a React component, which requires camelCase input)
  - [ ] Curlies mix properly with other pages styles (e.g. the page "shadow" option from the toolbar)
  - [ ] `\page` on the first line of a document will *not* create a page break, and is counted as page 1
- [ ] Verify old features have not broken
  - [ ] `\page` works without values.
  - [ ] Highlighting of `\page` still works
  - [ ] Page number display is accurate
  - [ ] Jumping to source/brew still works in every case (including Legacy)
- [ ] Test for edge cases / try to break things
  - [ ] Ensure CSS Properties ( --FOO ) can be set.
- [ ] Identify opportunities for simplification and refactoring
- [ ] Check for code legibility and appropriate comments

**Comment by 5e-Cleric** [comment]
`/page` in the first line in legacy creates a page, not in v3, i assume this is intended

**Comment by 5e-Cleric** [comment]
Custom properties `--foo` are not being set, furthermore, a custom property with the same name as a norrmal property is added as that normal property, so `--border` is added as `border`.

**Comment by calculuschild** [comment]
> Right now, when setting style for the first page, spaces are being allowed between page break and style inject, but not for other pages. Additionally, these spaces, while they don't break the regex for the inject, do break the syntax highlight.

I see what is going on here...  First page is the only one that might go into the Marked parser without being filtered through the page split regex. Ok. So to align with the other injector behavior, I should allow spaces in every case.

> Also, class or id injection doesn't work for either ( \page{#myPageId,myPageClass}

Good catch; somehow I deleted the class part when doing merges. Will put that back in. ID's are intentionally left out, else it would mess with page selection on the toolbar, etc.

> /page in the first line in legacy creates a page, not in v3, i assume this is intended

Correct. This PR only applies to V3.

> Custom properties --foo are not being set, furthermore, a custom property with the same name as a norrmal property is added as that normal property, so --border is added as border.

I see what happened. When I added the camelCase conversion it also converts `--my-custom-variable` to `myCustomVariable`. I will tweak that line to exclude converting properties starting with `--`.

**Comment by calculuschild** [comment]
@5e-Cleric I have fixed all of these cases you reported. Can you confirm?

**Comment by 5e-Cleric** [comment]
I don't think it is a problem, but the order of attributes is different, can we fix that?

image

**Comment by calculuschild** [comment]
Can you clarify what steps result in the different order?

As for testing, I'm not fully sure how we would go about testing React output. I might put it off for later unless you know of a good strategy.

**Comment by 5e-Cleric** [comment]
> Can you clarify what steps result in the different order?

Making the same inject in page 1 vs page 2, with some text in the middle, not sure either.

**Comment by calculuschild** [comment]
> Making the same inject in page 1 vs page 2, with some text in the middle, not sure either.

Interesting. I'm not sure how much control we have over the order React chooses here. Given that Page 1 is the only different one, my best guess is it has something to do with the very first "empty" render before any pages have been split and parsed, But this doesn't impact the behavior of the document at all so I don't know if it's worth hunting down.

I think I will merge this and we can decide if its worth the effort to clean up the ordering later.
